### PR TITLE
Alicja legend added 

### PR DESCRIPTION
**Pull request by AliceBalfanz**
Added the possibility to retrieve a legend as png for a certain variable. The user can choose from colormaps provided by the matplotlib library and also additionally  set the value ranges of the legend. if no colormap or range is chosen, the default settings are displayed.

### CODE REVIEW

File: xcube_server/app.py [hunk 1]
```diff
@@ -56,6 +56,7 @@ def new_application(name: str=DEFAULT_NAME):
         (prefix + url_pattern('/datasets'), GetDatasetsJsonHandler),
         (prefix + url_pattern('/variables/{{ds_name}}'), GetVariablesJsonHandler),
         (prefix + url_pattern('/coords/{{ds_name}}/{{dim_name}}'), GetCoordinatesJsonHandler),
+        (prefix + url_pattern('/{{ds_name}}/{{var_name}}/legend.png'), GetLegendHandler),
```
Comment (forman) [inline xcube_server/app.py]: URL style contradicts the one of other URLs in this API. Please be consistent. I suggest naming the resource first

```suggestion
        (prefix + url_pattern('/legend/{{ds_name}}/{{var_name}}.png'), GetLegendHandler),
```

File: xcube_server/context.py [hunk 2]
```diff
@@ -242,6 +242,14 @@ def get_dataset(self, ds_name: str) -> xr.Dataset:
 
         return ds
 
+    def get_legend_lable(self, ds_name: str, var_name: str):
```
Comment (forman) [inline xcube_server/context.py]: Naming style:

```suggestion
    def get_legend_label(self, ds_name: str, var_name: str):
```

File: xcube_server/context.py [hunk 3]
```diff
@@ -242,6 +242,14 @@ def get_dataset(self, ds_name: str) -> xr.Dataset:
...
+        dataset = self.get_dataset(ds_name)
+        if var_name in dataset:
+            ds = self.get_dataset(ds_name)
+            unit = ds[var_name].units
```
Comment (forman) [inline xcube_server/context.py]: Naming:

```suggestion
            units = ds[var_name].units
            return units
```

File: xcube_server/controllers/tiles.py [hunk 4]
```diff
@@ -1,15 +1,22 @@
 import time
 from typing import Dict, Any
 
+import matplotlib
+import matplotlib.cm as cm
+import matplotlib.pyplot as plt
```
Comment (forman) [inline xcube_server/controllers/tiles.py]: Wrong API,  `matplotlib.pyplot` is a stateful module. Use it in interactive envs only.

File: xcube_server/controllers/tiles.py [hunk 5]
```diff
@@ -117,6 +124,39 @@ def array_image_id_factory(level):
...
+        cmap_cbar = cmap_cbar or default_cmap_cbar
+        cmap_vmin = cmap_vmin or default_cmap_vmin
+        cmap_vmax = cmap_vmax or default_cmap_vmax
+
+    try:
+        cmap = cm.get_cmap(cmap_cbar)
+    except ValueError:
+        raise ServiceResourceNotFoundError(f"color bar {cmap_cbar} not found")
+
+    fig = plt.figure(figsize=(1, 5))
```
Comment (forman) [inline xcube_server/controllers/tiles.py]: See above, instantiate figures using `Figure` constructor,

File: xcube_server/controllers/tiles.py [hunk 6]
```diff
@@ -117,6 +124,39 @@ def array_image_id_factory(level):
...
+    norm = matplotlib.colors.Normalize(vmin=cmap_vmin, vmax=cmap_vmax)
+    image_legend = matplotlib.colorbar.ColorbarBase(ax1, cmap=cmap,
+                                                    norm=norm, orientation='vertical')
+    image_legend.set_label(ctx.get_legend_lable(ds_name, var_name))
+    fig.patch.set_facecolor('white')
+    fig.patch.set_alpha(0.0)
+    fig.tight_layout()
+
+    buffer = io.BytesIO()
+    plt.savefig(buffer, format='png')
```
Comment (forman) [inline xcube_server/controllers/tiles.py]: See above. There should be an alternative to `plt.savefig()`.

File: xcube_server/controllers/tiles.py [hunk 7]
```diff
@@ -117,6 +124,39 @@ def array_image_id_factory(level):
...
+        cmap = cm.get_cmap(cmap_cbar)
+    except ValueError:
+        raise ServiceResourceNotFoundError(f"color bar {cmap_cbar} not found")
+
+    fig = plt.figure(figsize=(1, 5))
+    ax1 = fig.add_subplot(1, 1, 1)
+    norm = matplotlib.colors.Normalize(vmin=cmap_vmin, vmax=cmap_vmax)
+    image_legend = matplotlib.colorbar.ColorbarBase(ax1, cmap=cmap,
+                                                    norm=norm, orientation='vertical')
+    image_legend.set_label(ctx.get_legend_lable(ds_name, var_name))
```
Comment (forman) [inline xcube_server/controllers/tiles.py]: Note that `ctx.get_legend_lable()` can return `None`. Does `image_legend.set_label()` accept that?

**Comment by AliceBalfanz** [comment]
Adjusted the code based on comments of forman, and can be reviewed once more. @forman

File: xcube_server/__init__.py [hunk 8]
```diff
@@ -20,5 +20,5 @@
 # SOFTWARE.
 
 # Version string (PEP440-compatible), e.g. "0.8.0", "0.8.0.dev1", "0.8.0rc1", "0.8.0rc1.dev1"
-__version__ = '0.1.0.dev4'
+__version__ = '0.2.0.dev4'
```
Comment (forman) [inline xcube_server/__init__.py]:
```suggestion
__version__ = '0.1.0.dev5'
```

File: xcube_server/controllers/tiles.py [hunk 9]
```diff
@@ -3,19 +3,18 @@
...
-from PIL import Image
 import io
+import matplotlib.figure
 
 from xcube_server.im import ImagePyramid, TransformArrayImage, ColorMappedRgbaImage, TileGrid
 from xcube_server.ne2 import NaturalEarth2Image
 from xcube_server.utils import compute_tile_grid
 from ..context import ServiceContext
-from ..defaults import TRACE_PERF
+from ..defaults import TRACE_PERF, DEFAULT_CMAP_W, DEFAULT_CMAP_H
```
Comment (forman) [inline xcube_server/controllers/tiles.py]:
```suggestion
from ..defaults import TRACE_PERF, DEFAULT_CMAP_WIDTH, DEFAULT_CMAP_HEIGHT
```

File: xcube_server/controllers/tiles.py [hunk 10]
```diff
@@ -3,19 +3,18 @@
...
-import matplotlib.pyplot as plt
 import matplotlib.colors
 import matplotlib.colorbar
 import numpy as np
 import xarray as xr
-from PIL import Image
 import io
+import matplotlib.figure
 
 from xcube_server.im import ImagePyramid, TransformArrayImage, ColorMappedRgbaImage, TileGrid
```
Comment (forman) [inline xcube_server/controllers/tiles.py]: Use relative paths

File: xcube_server/res/openapi.yml [hunk 11]
```diff
@@ -76,6 +76,28 @@ paths:
             application/json:
               schema:
                 $ref: '#/components/schemas/CoordinateVariable'
+  '/legend/{ds_name}/{var_name}.png':
+    get:
+      tags:
+        - tile
+      summary: Get legend
```
Comment (forman) [inline xcube_server/res/openapi.yml]:
```suggestion
      summary: Get legend for layer
```

File: xcube_server/res/openapi.yml [hunk 12]
```diff
@@ -76,6 +76,28 @@ paths:
...
+      operationId: getLegend
+      parameters:
+        - $ref: '#/components/parameters/dataset'
+        - $ref: '#/components/parameters/variable'
+      description: |
+        Get a legend of a given variable of a given dataset.
```
Comment (forman) [inline xcube_server/res/openapi.yml]: ```Get the legend for a layer of a given variable of a given dataset.```

File: xcube_server/__init__.py [hunk 13]
```diff
@@ -20,5 +20,5 @@
 # SOFTWARE.
 
 # Version string (PEP440-compatible), e.g. "0.8.0", "0.8.0.dev1", "0.8.0rc1", "0.8.0rc1.dev1"
-__version__ = '0.1.0.dev4'
+__version__ = '0.1.0.dev5'
```
Comment (TonioF) [inline xcube_server/__init__.py]: If you update the version, you should do it everywhere - also in the readme, in PKG-INFO, and the other places.

File: xcube_server/context.py [hunk 14]
```diff
@@ -242,6 +242,15 @@ def get_dataset(self, ds_name: str) -> xr.Dataset:
 
         return ds
```
Comment (TonioF) [inline xcube_server/context.py]: One blank line too many.

File: xcube_server/controllers/tiles.py [hunk 15]
```diff
@@ -117,6 +123,48 @@ def array_image_id_factory(level):
...
+    fig = matplotlib.figure.Figure(figsize=(cmap_w, cmap_h))
+    ax1 = fig.add_subplot(1, 1, 1)
+    norm = matplotlib.colors.Normalize(vmin=cmap_vmin, vmax=cmap_vmax)
+    image_legend = matplotlib.colorbar.ColorbarBase(ax1, cmap=cmap,
+                                                    norm=norm, orientation='vertical')
+
+    image_legend_label = ctx.get_legend_label(ds_name, var_name)
+    if image_legend_label is not None:
+        image_legend.set_label(image_legend_label)
+    # print('units not none')
```
Comment (TonioF) [inline xcube_server/controllers/tiles.py]: Remove commented line (or add comment why you want to keep it)

**Pull request merged**